library/spdm_common_lib/context_data.c

```diff
@@ -835,50 +1020,293 @@ return_status spdm_append_message_m(IN void *context, IN void *message,
+#if RECORD_TRANSCRIPT_DATA
 	return append_managed_buffer(
 		&spdm_session_info->session_transcript.message_k, message,
 		message_size);
+#else
+	{
+		spdm_context_t *spdm_context;
+		void *secured_message_context;
+		uint8 *cert_chain_buffer;
+		uintn cert_chain_buffer_size;
+		boolean result;
+		uint8 cert_chain_buffer_hash[MAX_HASH_SIZE];
+		uint32 hash_size;
+		boolean finished_key_ready;
+
+		spdm_context = context;
+		secured_message_context = spdm_get_secured_message_context_via_session_info (session_info);
+		finished_key_ready = spdm_secured_message_is_finished_key_ready(secured_message_context);
+
+		if (spdm_session_info->session_transcript.digest_context_th == NULL) {
+			if (!spdm_session_info->use_psk) {
+				if (is_requester) {
+					result = spdm_get_peer_cert_chain_buffer(
+						spdm_context, (void **)&cert_chain_buffer, &cert_chain_buffer_size);
+				} else {
+					result = spdm_get_local_cert_chain_buffer(
+						spdm_context, (void **)&cert_chain_buffer, &cert_chain_buffer_size);
+				}
+				if (!result) {
+					return FALSE;
+				}
+				hash_size = spdm_get_hash_size(
+					spdm_context->connection_info.algorithm.base_hash_algo);
+				spdm_hash_all(
+					spdm_context->connection_info.algorithm.base_hash_algo,
+					cert_chain_buffer, cert_chain_buffer_size,
+					cert_chain_buffer_hash);
+			}
+		}
+
+		//
+		// prepare digest_context_th
+		//
+		if (spdm_session_info->session_transcript.digest_context_th == NULL) {
+			spdm_session_info->session_transcript.digest_context_th = spdm_hash_new (
+				spdm_context->connection_info.algorithm.base_hash_algo);
+			spdm_hash_init (spdm_context->connection_info.algorithm.base_hash_algo,
+				spdm_session_info->session_transcript.digest_context_th);
+			spdm_hash_update (spdm_context->connection_info.algorithm.base_hash_algo,
+				spdm_session_info->session_transcript.digest_context_th,
+				get_managed_buffer(&spdm_context->transcript.message_a),
+				get_managed_buffer_size(&spdm_context->transcript.message_a));
+			append_managed_buffer(
+				&spdm_session_info->session_transcript.temp_message_k,
+				get_managed_buffer(&spdm_context->transcript.message_a),
+				get_managed_buffer_size(&spdm_context->transcript.message_a));
+			if (!spdm_session_info->use_psk) {
+				spdm_hash_update (spdm_context->connection_info.algorithm.base_hash_algo,
+					spdm_session_info->session_transcript.digest_context_th,
+					cert_chain_buffer_hash, hash_size);
+				append_managed_buffer(
+					&spdm_session_info->session_transcript.temp_message_k,
+					cert_chain_buffer_hash, hash_size);
+			}
+		}
+		spdm_hash_update (spdm_context->connection_info.algorithm.base_hash_algo,
+			spdm_session_info->session_transcript.digest_context_th, message, message_size);
+		if (!finished_key_ready) {
+			append_managed_buffer(
+				&spdm_session_info->session_transcript.temp_message_k, message, message_size);
+		}
+
+		//
+		// append message only if finished_key is NOT ready.

```

Comment:
I'm a little confused as to why we both manage temporary transcripts, as well as hashes here.

Is the idea that, we keep transcripts for all session establishments scenarios ONLY until the finished key is created?
And once we have finished key, we can then create HMAC's. So that all further updates are added to HMAC context, rather than transcripts themselves?

Reply:
that is right. before finished_key is ready, I am not able to HMAC.

Reply:
that is right. before finished_key is ready, I am not able to HMAC.

Reply:
Fixed